doc/template_definition.md

```diff
@@ -81,8 +124,79 @@ method.getBody().insertBegin(injectedCode);
 
 ```
 
-Kinds of templating
--------------------
+Definition of templates using `PatternBuilder`
+--------------------------------------------------
+
+The body of method `CheckBoundTemplate#statement` below defines a Spoon template. 
+
+```java
+public class CheckBoundTemplate /*it doesn't matter what it extends or implements*/ {
+	// it doesn't matter what other type members are in template class
+	// it doesn't matter what is the name of the method whose body will be used as template
+	public void statement(Collection<?> _col_) {
+		if (_col_.size() > 10)
+			throw new OutOfBoundException();
+	}
+}
+```
+
+The code, which creates a Spoon template using `PatternBuilder` looks like this:
+```java
+Pattern t = PatternBuilder.create(factory,
+	//defines template class.
+	CheckBoundTemplate.class, 
+	//defines which part of template class will be used as template model
+	model -> model.setBodyOfMethod("statement"))
+	//tells builder that all variables defined out of scope of template model (body of the method)
+	//are considered as template parameters
+	.configureAutomaticParameters()
+	//builds an instance of Pattern
+	.build();
+```
+
+
+This template specifies a
+statements (all statmenets of body of method `statement`) that is a precondition to check that a list 
+is smaller than a certain size. This piece of code will be injected at the 
+beginning of all methods dealing with size-bounded lists. This template has 
+one single template parameter called `_col_`. 
+In this case, the template parameter value is meant to be an expression (`CtExpression`) 
+that returns a Collection.
+
+The template source is 
+well-typed, compiles, but the binary code of the template is usually thrown away
+and only spoon model (Abstract syntax tree) of source code is used to generated new code
+or to search for matching code.
+
+Generating of code using Spoon template
+-------------
+
+The code at the end of this page shows how to use such spoon template.
+One takes a spoon template, defines the template parameters, 
+and then one calls the template engine. In last line, the bound check 
+is injected at the beginning of a method body. 
+
+Since the template is given the first method parameter which is in the 
+scope of the insertion location, the generated code is guaranteed to compile. 
+The Java compiler ensures that the template compiles with a given scope, the 
+developer is responsible for checking that the scope where she uses 
+template-generated code is consistent with the template scope.
+
+```java
+Pattern t = PatternBuilder.create(...building a template. See code above...);
+// creating a holder of parameters
+Map<String, Object> parameters = new HashMap<>();
+parameters.put("_col_", createVariableAccess(method.getParameters().get(0))); 
+
+// getting the final AST
+CtStatement injectedCode = t.substituteSingle(factory, CtStatement.class, parameters);
```

Comment:
Why do you need to specify the `CtElement` type in `substituteSingle`: you know that `t` refers to a `CtStatement` so you only can substitute here to another `CtStatement`, no?

In other word: what would happen if I type: 

`CtMethod injectedCode = t.substituteSingle(factory, CtMethod.class, parameters);`

Reply:
It will fail because CtStatement cannot be converted to CtMethod. But if you would use
```java
CtBlock injectedCode = t.substituteSingle(factory, CtBlock.class, parameters);
```
then you would get a CtBlock which contains your generated statements. Usually the definition of required type is useless, but I already seen cases when the definition is absolutely necessary - it is case of legacy templates too. So it is part of that API now.

Reply:
ok